merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java

```diff
@@ -383,20 +332,107 @@ public boolean isTaskComplete(final TaskId task) {
+        return taskInfo -> { left.accept(taskInfo); right.accept(taskInfo); };
+      }
+
+      public Consumer<TaskInfo> atom(final Event ev) {
+        return taskInfo -> {
+          // Identify activities.
+          ev.extract(this.activityTopic)
+            .ifPresent(directiveId -> taskInfo.taskToPlannedDirective.put(ev.provenance().id(), directiveId));
+
+          for (final var topic : this.missionModel.getTopics()) {
+            // Identify activity inputs.
+            extractInput(topic, ev, taskInfo);
+
+            // Identify activity outputs.
+            extractOutput(topic, ev, taskInfo);
+          }
+        };
+      }
+
+      private static <T>
+      void extractInput(final MissionModel.SerializableTopic<T> topic, final Event ev, final TaskInfo taskInfo) {
+        if (!topic.name().startsWith("ActivityType.Input.")) return;
+
+        ev.extract(topic.topic()).ifPresent(input -> {
+          final var activityType = topic.name().substring("ActivityType.Input.".length());
+
+          taskInfo.input.put(
+              ev.provenance().id(),
+              new SerializedActivity(activityType, topic.serializer().apply(input).asMap().orElseThrow()));
+        });
+      }
+
+      private static <T>
+      void extractOutput(final MissionModel.SerializableTopic<T> topic, final Event ev, final TaskInfo taskInfo) {
+        if (!topic.name().startsWith("ActivityType.Output.")) return;
+
+        ev.extract(topic.topic()).ifPresent(output -> {
+          taskInfo.output.put(
+              ev.provenance().id(),
+              topic.serializer().apply(output));
+        });
+      }
+    }
+  }
+
   /** Compute a set of results from the current state of simulation. */
   // TODO: Move result extraction out of the SimulationEngine.
   //   The Engine should only need to stream events of interest to a downstream consumer.
   //   The Engine cannot be cognizant of all downstream needs.
   // TODO: Whatever mechanism replaces `computeResults` also ought to replace `isTaskComplete`.
   // TODO: Produce results for all tasks, not just those that have completed.
   //   Planners need to be aware of failed or unfinished tasks.
-  public SimulationResults computeResults(
+  public static SimulationResults computeResults(
       final SimulationEngine engine,
       final Instant startTime,
       final Duration elapsedTime,
-      final Map<String, ActivityInstanceId> taskToPlannedDirective,
+      final Topic<ActivityInstanceId> activityTopic,
       final TemporalEventSource timeline,
-      final MissionModel<?> missionModel) {
+      final MissionModel<?> missionModel
+  ) {
+    // Collect per-task information from the event graph.
+    final var taskInfo = new TaskInfo();
+
+    for (final var point : timeline) {
+      if (!(point instanceof TemporalEventSource.TimePoint.Commit p)) continue;
+      if (!p.topics().contains(activityTopic)) continue;
```

Comment:
Another subtle point that I'll call out - this statement assumes that input and output attributes are only emitted during Commits that also include events on `activityTopic`. I'm having some trouble proving to myself whether or not this assumption is true.

`emitAndThen` emits to `activityTopic`, and then yields - I _think_ this means that any subsequent steps won't be in the same Commit. Certainly I wouldn't expect the output attributes to be in that Commit.

Reply:
> this statement assumes that input and output attributes are only emitted during Commits that also include events on `activityTopic`

Ah, I goofed -- originally this loop only handled `activityTopic` events, and I wanted to avoid processing commits that were definitely not of interest. This guard should simply be removed now.

For activities spawned by directive (i.e. they were in the input plan), technically the start attributes event would be in the same commit as the directive association event. This would not be trued for child activities (which aren't associated with any directive), and would anyway be almost never true for end attributes.

> `emitAndThen` emits to `activityTopic`, and then yields

No, it doesn't yield -- it emits and then immediately runs the remainder of the task (i.e. the task being wrapped).

Reply:
> This guard should simply be removed now.

I've removed this guard from the commit and re-pushed.

Reply:
Looks great, thanks!